File: frescophotoview/src/main/java/com/zml/frescophotoview/DefaultGestureHandler.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.zml.frescophotoview;

import android.graphics.PointF;
import android.view.GestureDetector;
import android.view.MotionEvent;

import static com.zml.frescophotoview.ZoomableControllerImp.LIMIT_ALL;

/**
 * Tap gesture listener for double tap to zoom / unzoom and double-tap-and-drag to zoom.
 */
public class DefaultGestureHandler extends GestureDetector.SimpleOnGestureListener {
    private final FrescoPhotoView mFrescoPhotoView;
    private final PointF mDoubleTapViewPoint = new PointF();

    public DefaultGestureHandler(FrescoPhotoView frescoPhotoView) {
        mFrescoPhotoView = frescoPhotoView;
    }

    @Override
    public void onLongPress(MotionEvent e) {
        super.onLongPress(e);
    }

    @Override
    public boolean onSingleTapConfirmed(MotionEvent e) {
        return super.onSingleTapConfirmed(e);
    }

    @Override
    public boolean onDoubleTapEvent(MotionEvent e) {
        final ZoomableControllerImp zc =
                (ZoomableControllerImp) mFrescoPhotoView.getZoomableController();
        PointF vp = new PointF(e.getX(), e.getY());
        switch (e.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                mDoubleTapViewPoint.set(vp);
                break;
            case MotionEvent.ACTION_UP:
                ScaleFactorRetriever fetcher = zc.getScaleFactorFetcher();
                fetcher.nextFactor(zc.getScaleFactor());
                zc.setDisableGesture(true);
                zc.zoomToPoint(
                        fetcher.nextFactor(zc.getScaleFactor()), vp, LIMIT_ALL, zc.getAnimationDuration(), () -> zc.setDisableGesture(false), null);
                break;
        }
        return true;
    }


    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        final ZoomableControllerImp zc =
                (ZoomableControllerImp) mFrescoPhotoView.getZoomableController();
        if (zc.getScaleFactor() > zc.getMinScaleFactor()) {
            zc.fling((int) velocityX, (int) velocityY);
        }
        return false;
    }
}

File: frescophotoview/src/main/java/com/zml/frescophotoview/DefaultGestureHandler.kt
package com.zml.frescophotoview

import android.graphics.PointF
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent

/**
 * 默认的行为处理双击 fling动作
 */
open class DefaultGestureHandler(private val zoomableControllerImp: ZoomableControllerImp) :
    SimpleOnGestureListener() {
    private val mDoubleTapViewPoint = PointF()
    override fun onLongPress(e: MotionEvent) {
        super.onLongPress(e)
    }

    override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
        return super.onSingleTapConfirmed(e)
    }

    override fun onDoubleTapEvent(e: MotionEvent): Boolean {
        val zc = zoomableControllerImp
        val vp = PointF(e.x, e.y)
        when (e.actionMasked) {
            MotionEvent.ACTION_DOWN -> mDoubleTapViewPoint.set(vp)
            MotionEvent.ACTION_UP -> {
                val fetcher = zc.scaleFactorFetcher
                fetcher.nextFactor(zc.scaleFactor)
                zc.isDisableGesture = true
                zc.zoomToPoint(
                    fetcher.nextFactor(zc.scaleFactor),
                    vp,
                    { zc.isDisableGesture = false },
                    null
                )
            }
        }
        return true
    }

    override fun onFling(
        e1: MotionEvent,
        e2: MotionEvent,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        val zc = zoomableControllerImp
        if (zc.scaleFactor > zc.minScaleFactor) {
            zc.fling(velocityX.toInt(), velocityY.toInt())
        }
        return false
    }
}